pkg/cloudfoundry/YamlUtil.go

```diff
@@ -0,0 +1,246 @@
+package cloudfoundry
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/SAP/jenkins-library/pkg/log"
+	"gopkg.in/yaml.v2"
+	"io"
+	"io/ioutil"
+	"os"
+	"reflect"
+	"regexp"
+	"strings"
+)
+
+type fUtils interface {
+	FileRead(name string) ([]byte, error)
+	FileWrite(name string, data []byte, mode os.FileMode) error
+}
+
+type fileUtils struct{}
+
+func (s *fileUtils) FileRead(name string) ([]byte, error) {
+	return ioutil.ReadFile(name)
+}
+
+func (s *fileUtils) FileWrite(name string, data []byte, mode os.FileMode) error {
+	return ioutil.WriteFile(name, data, mode)
+}
+
+var _fileUtils fUtils
+
+var _stat = os.Stat
+var _traverse = traverse
+
+// Substitute ...
+func Substitute(ymlFile string, replacements map[string]interface{}, replacementsFiles []string) (bool, error) {
```

Comment:
Just an idea: Instead of the global _fileUtils pointer, you could have a public and a private version of Substitute():

```golang
func Substitute(ymlFile string, replacements map[string]interface{}, replacementsFiles []string) (bool, error) {
	return substitute(ymlFile, replacements, replacementsFiles, &piperutils.Files{})
}

func substitute(ymlFile string, replacements map[string]interface{}, replacementsFiles []string, fileUtils fUtils) (bool, error) {
...
}
```

You could then test the private version without messing with global variables. There would not be any need for the fileUtils struct (line 21 to 31).

Reply:
@stippi2 

> you could have a public and a private version of Substitute()

I implemented that [here](https://github.com/marcusholl/jenkins-library/commit/18e6582ceb2da7de0238869ab1f6c53d32dd4d9b).

>without messing with global variables

I don't see a disadvantage with having a global variable. We would have problems when we would execute tests in parallel with different mocks. But that I don't do. We have to make the signatures of other funcs broader (forwarding the utils) rather than using the global var. I like meaningful signatures getting handed over the entities they need. The file utils are present only due to testing purposes and basically more the less auxiliar instead of calling the file-like methods directly when we wouldn't have the tests.

>There would not be any need for the fileUtils struct (line 21 to 31)

This I don't understand - at least not in the context of global variables. We have to provide something implementing the interface for the free wild. But I guess we can change that to the already existing `piperutils.FileUtils`.